File: src/exoticatechnologies/modifications/exotics/Exotic.java
package exoticatechnologies.modifications.exotics;

import com.fs.starfarer.api.Global;
import com.fs.starfarer.api.campaign.CampaignFleetAPI;
import com.fs.starfarer.api.campaign.SpecialItemData;
import com.fs.starfarer.api.campaign.econ.MarketAPI;
import com.fs.starfarer.api.combat.MutableShipStatsAPI;
import com.fs.starfarer.api.combat.ShipAPI;
import com.fs.starfarer.api.fleet.FleetMemberAPI;
import com.fs.starfarer.api.ui.TooltipMakerAPI;
import com.fs.starfarer.api.ui.UIComponentAPI;
import exoticatechnologies.modifications.Modification;
import exoticatechnologies.modifications.ShipModifications;
import exoticatechnologies.util.StringUtils;
import lombok.Getter;
import lombok.Setter;
import org.json.JSONObject;

import java.awt.*;
import java.util.HashMap;
import java.util.Map;

public abstract class Exotic extends Modification {
    public static final String ITEM = "et_exotic";
    @Getter
    @Setter
    protected String description;
    @Getter
    @Setter
    protected String tooltip;

    public Exotic(String key, JSONObject settings) {
        super(key, settings);
    }

    public static Exotic get(String exoticKey) {
        return ExoticsHandler.EXOTICS.get(exoticKey);
    }

    public abstract Color getColor();

    public String getTextDescription() {
        return getDescription() + "\n\n" + getTooltip();
    }

    public String getIcon() {
        return "graphics/icons/exotics/" + getKey().toLowerCase() + ".png";
    }

    public String getBuffId() {
        return "ET_" + getKey();
    }

    protected boolean isNPC(FleetMemberAPI fm) {
        return fm.getFleetData() == null
                || fm.getFleetData().getFleet() == null
                || !fm.getFleetData().getFleet().equals(Global.getSector().getPlayerFleet());
    }

    public void onInstall(FleetMemberAPI fm) {

    }

    public void onDestroy(FleetMemberAPI fm) {

    }

    public boolean canAfford(CampaignFleetAPI fleet, MarketAPI market) {
        return false;
    }

    public boolean removeItemsFromFleet(CampaignFleetAPI fleet, FleetMemberAPI fm, MarketAPI market) {
        return false;
    }

    public void printDescriptionToTooltip(FleetMemberAPI fm, TooltipMakerAPI tooltip) {
        StringUtils.getTranslation(this.getKey(), "description")
                .addToTooltip(tooltip);
    }

    public void printStatInfoToTooltip(FleetMemberAPI fm, TooltipMakerAPI tooltip) {
        StringUtils.getTranslation(this.getKey(), "longDescription")
                .addToTooltip(tooltip);
    }

    public void printDescriptionToTooltip(TooltipMakerAPI tooltip, FleetMemberAPI member) {
        StringUtils.getTranslation(this.getKey(), "description")
                .addToTooltip(tooltip);
    }

    public void printStatInfoToTooltip(TooltipMakerAPI tooltip, FleetMemberAPI member) {
        StringUtils.getTranslation(this.getKey(), "longDescription")
                .addToTooltip(tooltip);
    }

    public Map<String, Float> getResourceCostMap(FleetMemberAPI fm, ShipModifications mods, MarketAPI market) {
        return new HashMap<>();
    }


    /**
     * extra bandwidth allowed to be installed.
     *
     * @param fm
     * @param es
     * @return
     */
    public float getExtraBandwidthPurchaseable(FleetMemberAPI fm, ShipModifications es) {
        return 0f;
    }

    /**
     * extra bandwidth added directly to ship.
     *
     * @param fm
     * @param es
     * @return
     */
    public float getExtraBandwidth(FleetMemberAPI fm, ShipModifications es) {
        return 0f;
    }

    public void applyExoticToStats(FleetMemberAPI fm, MutableShipStatsAPI stats, float bandwidth, String id) {

    }

    public void applyExoticToShip(FleetMemberAPI fm, ShipAPI ship, float bandwidth, String id) {

    }

    public void advanceInCombatUnpaused(ShipAPI ship, float amount, float bandwidth) {

    }

    public void advanceInCombatAlways(ShipAPI ship, float bandwidth) {

    }

    public void advanceInCampaign(FleetMemberAPI member, ShipModifications mods, Float amount) {

    }

    public boolean canDropFromFleets() {
        return true;
    }

    public float getSalvageChance(float chanceMult) {
        return 0.2f * chanceMult;
    }

    public SpecialItemData getNewSpecialItemData() {
        return new SpecialItemData(Exotic.ITEM, this.getKey());
    }

    public abstract void modifyToolTip(TooltipMakerAPI tooltip, UIComponentAPI title, FleetMemberAPI fm, ShipModifications systems, boolean expand);
}

File: src/exoticatechnologies/modifications/exotics/Exotic.kt
package exoticatechnologies.modifications.exotics

import com.fs.starfarer.api.Global
import com.fs.starfarer.api.campaign.CampaignFleetAPI
import com.fs.starfarer.api.campaign.SpecialItemData
import com.fs.starfarer.api.campaign.econ.MarketAPI
import com.fs.starfarer.api.combat.MutableShipStatsAPI
import com.fs.starfarer.api.combat.ShipAPI
import com.fs.starfarer.api.fleet.FleetMemberAPI
import com.fs.starfarer.api.ui.TooltipMakerAPI
import com.fs.starfarer.api.ui.UIComponentAPI
import exoticatechnologies.modifications.Modification
import exoticatechnologies.modifications.ShipModifications
import exoticatechnologies.modifications.conditions.toList
import exoticatechnologies.modifications.exotics.types.ExoticType
import exoticatechnologies.ui.impl.shop.exotics.methods.ExoticMethod
import exoticatechnologies.util.StringUtils
import org.json.JSONObject
import java.util.*

abstract class Exotic(key: String, settings: JSONObject) : Modification(key, settings) {
    var loreDescription: String? = null
    open val textDescription: String?
        get() = "$loreDescription\n\n$description"

    public override var icon: String = settings.optString("icon", key)
    val buffId: String
        get() = "ET_$key"
    var allowedMethods: List<String>? = null
    var blockedMethods: List<String>? = null

    init {
        if (settings.has("allowedMethods")) {
            allowedMethods = settings.optJSONArray("allowedMethods").toList()
        }

        if (settings.has("blockedMethods")) {
            blockedMethods = settings.optJSONArray("blockedMethods").toList()
        }
    }

    protected fun isNPC(fm: FleetMemberAPI): Boolean {
        return fm.fleetData == null || fm.fleetData.fleet == null || fm.fleetData.fleet != Global.getSector().playerFleet
    }

    open fun onInstall(member: FleetMemberAPI) {}
    open fun onDestroy(member: FleetMemberAPI) {}
    open fun canAfford(fleet: CampaignFleetAPI, market: MarketAPI?): Boolean {
        return false
    }

    open fun getGenerationChanceMult(member: FleetMemberAPI): Float {
        return 1f
    }

    open fun countsTowardsExoticLimit(member: FleetMemberAPI): Boolean {
        return true
    }

    open fun removeItemsFromFleet(fleet: CampaignFleetAPI, member: FleetMemberAPI, market: MarketAPI?): Boolean {
        return false
    }

    fun printDescriptionToTooltip(fm: FleetMemberAPI, tooltip: TooltipMakerAPI) {
        StringUtils.getTranslation(key, "description")
            .addToTooltip(tooltip)
    }

    fun printStatInfoToTooltip(fm: FleetMemberAPI, tooltip: TooltipMakerAPI) {
        StringUtils.getTranslation(key, "longDescription")
            .addToTooltip(tooltip)
    }

    fun printDescriptionToTooltip(tooltip: TooltipMakerAPI, member: FleetMemberAPI) {
        StringUtils.getTranslation(key, "description")
            .addToTooltip(tooltip)
    }

    fun printStatInfoToTooltip(tooltip: TooltipMakerAPI, member: FleetMemberAPI) {
        StringUtils.getTranslation(key, "longDescription")
            .addToTooltip(tooltip)
    }

    open fun getResourceCostMap(
        fm: FleetMemberAPI,
        mods: ShipModifications,
        market: MarketAPI?
    ): MutableMap<String, Float> {
        return HashMap()
    }

    /**
     * extra bandwidth allowed to be installed.
     *
     * @param member
     * @param mods
     * @param exoticData
     * @return
     */
    fun getExtraBandwidthPurchaseable(member: FleetMemberAPI, mods: ShipModifications, exoticData: ExoticData?): Float {
        return 0f
    }

    /**
     * extra bandwidth added directly to ship.
     *
     * @param member
     * @param mods
     * @param exoticData
     * @return
     */
    open fun getExtraBandwidth(member: FleetMemberAPI, mods: ShipModifications, exoticData: ExoticData?): Float {
        return 0f
    }

    open fun applyExoticToStats(
        id: String,
        stats: MutableShipStatsAPI,
        member: FleetMemberAPI,
        mods: ShipModifications,
        exoticData: ExoticData
    ) {
    }

    open fun applyToShip(
        id: String,
        member: FleetMemberAPI,
        ship: ShipAPI,
        mods: ShipModifications,
        exoticData: ExoticData
    ) {
    }

    open fun applyToFighters(member: FleetMemberAPI, ship: ShipAPI, fighter: ShipAPI, mods: ShipModifications) {
    }

    open fun advanceInCombatUnpaused(
        ship: ShipAPI,
        amount: Float,
        member: FleetMemberAPI,
        mods: ShipModifications,
        exoticData: ExoticData
    ) {
    }

    open fun advanceInCombatAlways(
        ship: ShipAPI,
        member: FleetMemberAPI,
        mods: ShipModifications,
        exoticData: ExoticData
    ) {
    }

    open fun advanceInCampaign(
        member: FleetMemberAPI,
        mods: ShipModifications,
        amount: Float,
        exoticData: ExoticData
    ) {
    }

    open fun getSalvageChance(chanceMult: Float): Float {
        return 0.2f * chanceMult
    }

    open fun canUseMethod(member: FleetMemberAPI?, mods: ShipModifications, method: ExoticMethod): Boolean {
        if (blockedMethods?.contains(method.key) == true) {
            return false
        }

        if (allowedMethods?.contains(method.key) == false) {
            return false
        }

        return true
    }

    open fun canUseExoticType(type: ExoticType): Boolean {
        return true
    }

    open fun getPositiveMult(member: FleetMemberAPI, mods: ShipModifications, exoticType: ExoticType?): Float {
        return exoticType?.getPositiveMult(member, mods) ?: 1f
    }

    fun getPositiveMult(member: FleetMemberAPI, mods: ShipModifications, exoticData: ExoticData?): Float {
        return getPositiveMult(member, mods, exoticData?.type)
    }

    open fun getNegativeMult(member: FleetMemberAPI, mods: ShipModifications, exoticType: ExoticType?): Float {
        return exoticType?.getNegativeMult(member, mods) ?: 1f
    }

    fun getNegativeMult(member: FleetMemberAPI, mods: ShipModifications, exoticData: ExoticData?): Float {
        return getNegativeMult(member, mods, exoticData?.type)
    }

    fun getNewSpecialItemData(exoticType: ExoticType): SpecialItemData {
        return SpecialItemData(ITEM, String.format("%s,%s", key, exoticType.nameKey))
    }

    val newSpecialItemData: SpecialItemData
        get() = SpecialItemData(ITEM, key)

    abstract fun modifyToolTip(
        tooltip: TooltipMakerAPI,
        title: UIComponentAPI,
        member: FleetMemberAPI,
        mods: ShipModifications,
        exoticData: ExoticData,
        expand: Boolean
    )

    companion object {
        const val ITEM = "et_exotic"
        operator fun get(exoticKey: String?): Exotic? {
            return ExoticsHandler.EXOTICS[exoticKey]
        }
    }
}